README.md
# HTTP Reply Component

## Table of Contents

* [Description](#description)
* [Actions](#actions)
  * [Reply](#reply)
  * [Reply With Attachment](#reply-with-attachment)

### Description

The component replies with messages to the client requested a webhook.

This component takes the incoming message body and applies the configured JSONata transformation on it, if present, and return a message back to the client requested a webhook of a given flow.

## Actions

### Reply

#### Configuration Fields
* **Custom HTTP Headers** - (string, optional): Provides with possibility to set additional headers separated by comma (e.g `Content-Language, User-Agent`)

#### Input Metadata
* **Content Type (Defaults to 'application/json')** - (string, optional, defaults to `application/json`): Header value tells the client what the content type of the returned content actually is.
* **Response Body** - (string/Object, required): Body to send as the response
* **Response Status Code** - (number, optional, defaults to `200`): Integer number between `200` and `999` (more info about status codes in [rfc7231](https://datatracker.ietf.org/doc/html/rfc7231#section-6) standart)

If provided `Custom HTTP Headers` there will be additional field:
* **customHeaders**, contains:
  * **Header <header name provided in "Custom HTTP Headers">** - you can provide value to your custom header here


#### Output Metadata
Same as `Input Metadata`

### Reply With Attachment

#### Configuration Fields
* **Custom HTTP Headers** - (string, optional): Provides with possibility to set additional headers separated by comma (e.g `Content-Language, User-Agent`)

#### Input Metadata
* **Content Type (Defaults to 'application/json')** - (string, optional, defaults to `application/json`): Header value tells the client what the content type of the returned content actually is.
* **Attachment URL** - (string, required): Link to file (on platform or external) that will be used as response
* **Response Status Code** - (number, optional, defaults to `200`): Integer number between `200` and `999` (more info about status codes in [rfc7231](https://datatracker.ietf.org/doc/html/rfc7231#section-6) standart)

If provided `Custom HTTP Headers` there will be additional field:
* **customHeaders**, contains:
  * **Header <header name provided in "Custom HTTP Headers">** - you can provide value to your custom header here


#### Output Metadata
Same as `Input Metadata`



CHANGELOG.md
## 1.3.0 (June 29, 2023)
* Added `Response Status Code` validation
* Updated README
* Get rid of vulnerabilities and unused libraries in dependencies

## 1.2.7 (November 04, 2022)
* Updated the sailor version to 2.7.1

## 1.2.6 (August 26, 2022)
* Updated @elasticio/maester-client to v4.0.3
* Updated @elasticio/component-commons-library to v3.0.2

## 1.2.5 (August 12, 2022)
* Updated @elasticio/maester-client to v4.0.2
* Updated @elasticio/component-commons-library to v3.0.1

## 1.2.4 (July 29, 2022)
* Change library for attachments processing (object-storage-client => maester-client)  
* Update component-commons-library to 3.0.0  
* Update maester-client to 4.0.0
* Update Sailor to 2.6.29

## 1.2.3 (April 14, 2022)
* Update Sailor version to 2.6.27
* Get rid of vulnerabilities in dependencies
* Add component pusher job to Circle.ci config

## 1.2.2 (November 12, 2021)
* Update sailor version to 2.6.26
* Fix a bug in the build with different error message

## 1.2.1 (May 7, 2021)
* Fixed sample retrieve bug in `Reply With Attachment`  action

## 1.2.0 (April 23, 2021)
* Added new action: `Reply With Attachment`

## 1.1.5 (January 28, 2021)
* Update sailor version to 2.6.24
* Fix a bug with memory leakage in Sailor

## 1.1.4 (January 15, 2021)
* Update sailor version to 2.6.23

## 1.1.3 (November 6, 2020)
* Update sailor version to 2.6.18

## 1.1.2 (October 21, 2020)
* Update sailor version to 2.6.17

## 1.1.1 (October 2, 2020)
* Annual audit of the component code to check if it exposes a sensitive data in the logs

## 1.1.0 (June 15, 2020)
* Make it possible to configure response status code

## 1.0.4 (May 19, 2020)
* Update sailor version to 2.6.7

## 1.0.3 (January 30, 2020)
* Update sailor version to 2.6.1

## 1.0.2 (December 07, 2019)
* Update sailor version to 2.5.1


lib/ReplyClient.js
const { messages } = require('elasticio-node');
const commons = require('@elastic.io/component-commons-library');
const utils = require('./utils');

const DEFAULT_CONTENT_TYPE = 'application/json';
const DEFAULT_STATUS_CODE = 200;
const HEADER_CONTENT_TYPE = 'Content-Type';
const HEADER_ROUTING_KEY = 'X-EIO-Routing-Key';
const HEADER_STATUS_CODE = 'x-eio-status-code';
const HEADER_OBJECT_STORAGE = 'x-ipaas-object-storage-id';

class ReplyClient {
  constructor(context, msg) {
    const {
      contentType = DEFAULT_CONTENT_TYPE,
      statusCode = DEFAULT_STATUS_CODE,
      customHeaders = {},
    } = msg.body;

    const {
      reply_to
    } = msg.headers;

    if (utils.isNumberNaN(statusCode) || Number(statusCode) < 200 || Number(statusCode) > 999) {
      const errorMessage = `"Response Status Code" must be valid number between 200 and 999, got ${this.statusCode}`;
      utils.logAndThrowError(context.logger, errorMessage);
    }

    this.reply_to = reply_to;
    this.contentType = contentType;
    this.statusCode = Number(statusCode);
    this.customHeaders = customHeaders;
    this.replyBody = {};
    this.logger = context.logger;
    this.context = context;
    this.msg = msg;
  }

  prepareReply() {
    const reply = messages.newMessageWithBody(this.replyBody);
    reply.headers[HEADER_ROUTING_KEY] = this.reply_to;
    reply.headers[HEADER_CONTENT_TYPE] = this.contentType;
    reply.headers[HEADER_STATUS_CODE] = this.statusCode;
    if (this.objectId) reply.headers[HEADER_OBJECT_STORAGE] = this.objectId;
    Object.assign(reply.headers, this.customHeaders);
    this.reply = reply;
  }

  async emitReply() {
    if (!this.reply_to) return;
    this.prepareReply();
    this.logger.info('Sending reply');
    await this.context.emit('data', this.reply);
  }

  async emitData() {
    this.logger.info('Emitting data...');
    await this.context.emit('data', messages.newMessageWithBody(this.msg.body));
  }

  async replyWithBody() {
    this.replyBody = this.msg.body.responseBody;
    if (!this.msg.body.responseBody) {
      this.logger.warn('Field "Response Body" on the message body was empty, we will reply with the whole message body');
      this.replyBody = this.msg.body;
    }
    await this.emitReply();
    await this.emitData();
  }

  async replyWithAttachment() {
    const { responseUrl } = this.msg.body;
    if (!responseUrl) {
      const errorMessage = '"Attachment URL" field can not be empty!';
      utils.logAndThrowError(this.logger, errorMessage);
    }
    const attachmentProcessor = new commons.AttachmentProcessor(utils.getUserAgent('component-commons-library'), this.msg.id);
    const objectStorage = utils.getObjectStorage();
    const getDataStream = async () => (await attachmentProcessor.getAttachment(responseUrl, 'stream')).data;
    this.objectId = await objectStorage.add(getDataStream);
    await this.emitReply();
    await this.emitData();
  }
}

module.exports.ReplyClient = ReplyClient;
